vowpalwabbit/learner.h

```diff
@@ -105,79 +115,98 @@ typedef void (*tend_example)(vw& all, void* d, example& ec);
   func_data end_pass_fd;
   func_data end_examples_fd;
   func_data finisher_fd;
+  test_example_data test_example_fd;
+  learner(){};  // Should only be able to construct a learner through init_learner function
 
 public:
   prediction_type::prediction_type_t pred_type;
   size_t weights; //this stores the number of "weight vectors" required by the learner.
   size_t increment;
+  bool is_multiline;  // Is this a single-line or multi-line reduction?
+
+  using end_fptr_type = void(*)(vw&, void*, void*);
+  using finish_fptr_type = void(*)(void*);
 
   //called once for each example.  Must work under reduction.
-  inline void learn(example& ec, size_t i=0)
-  { ec.ft_offset += (uint32_t)(increment*i);
-    learn_fd.learn_f(learn_fd.data, *learn_fd.base, ec);
-    ec.ft_offset -= (uint32_t)(increment*i);
+  inline void learn(E& ec, int32_t i = 0)
+  { assert((is_multiline && std::is_same<multi_ex, E>::value) ||
+      (!is_multiline && std::is_same<example, E>::value));  // sanity check under debug compile
+    adjust_offset(ec, increment, i);
+    learn_fd.learn_f(learn_fd.data, *learn_fd.base, (void*)&ec);
+    adjust_offset(ec, increment, -i);
```

Comment:
This is why you are doing int32_t.  It seems ok to use internally, but not in the external types.